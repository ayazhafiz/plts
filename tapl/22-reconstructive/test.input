lambda x. x;

lambda z. z 1;

plus = letrec plus =
  lambda m.
  lambda n.
    if iszero m
    then n
    else succ (plus (pred m) n)
  in plus;

not = lambda b. if b then false else true;

let twice = lambda f. lambda a. f(f(a)) in
  let three = twice (plus 1) 1 in
  let fls = twice (lambda n. n) false in
  if (not fls) then three else 1;

adder = let r = ref (lambda x. x)
  in (r := (lambda x. succ x); r);
(!adder)50;

let read = lambda x. x.a in
  let num = read {a=3, b=1} in
  read {a="abcdef", n=num};

getNoise = lambda animal.
           case animal of
               <dog=d> ==> d.noise
             | <cat=c> ==> c.noise
             | <whale=w> ==> w.noise;

getNoise <dog={noise="bark"}>;
/* NOTE :: this doesn't work because we're missing row variables! */
/* getNoise (<cat={noise="meow"}> as <cat:{noise:String}, whale:{noise:String}>); */
getNoise = getNoise; /* Notice the updated typings */

cell_fib = ref (lambda n. 0);
cell_fib := (
  lambda n.
    if iszero n then 0
    else if iszero (pred n) then n
    else plus ((!cell_fib) (pred n)) ((!cell_fib) (pred (pred n)))
);
fib = !cell_fib;

/* Oh, and type annotations still work too. You can mix annotated/unannotated, if you like :p */
addOrTimes = lambda m.
             lambda n: Float.
             lambda doAdd.
               if doAdd then (plusfloat m n) else (timesfloat m n);

addOrTimes 5.0 6.0 false;
