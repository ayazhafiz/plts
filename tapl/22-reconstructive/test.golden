(λ x. x): ??T0 -> ??T0
(λ z. (z <- 1)): Nat -> ??T2 -> ??T2
plus : Nat -> Nat -> Nat
not : Bool -> Bool
3: Nat
adder : Ref Nat -> Nat
51: Nat
abcdef: String
getNoise : <dog:{noise:??T39}, cat:{noise:??T39}, whale:{noise:??T39}> -> ??T39
bark: String
getNoise : <dog:{noise:String}, cat:{noise:String}, whale:{noise:String}> -> String
cell_fib : Ref ??T46 -> Nat
unit: Unit
fib : Nat -> Nat
addOrTimes : Float -> Float -> Bool -> Float
30.: Float
